RFC-0010-access-control-filters-masks.md

```diff
@@ -0,0 +1,205 @@
+    }
+```
+
+SystemAccessControl.java
+```java
+
+    /**
+     * Get row filters associated with the given table and identity.
+     * <p>
+     * Each filter must be a scalar SQL expression of boolean type over the columns in the table.
+     *
+     * @return a list of filters, or empty list if not applicable
+     */
+    default List<ViewExpression> getRowFilters(Identity identity, AccessControlContext context, CatalogSchemaTableName tableName)
+    {
+        return Collections.emptyList();
+    }
+
+    /**
+     * Bulk method for getting column masks for a subset of columns in a table.
+     * <p>
+     * Each mask must be a scalar SQL expression of a type coercible to the type of the column being masked. The expression
+     * must be written in terms of columns in the table.
+     *
+     * @return a mapping from columns to masks, or an empty map if not applicable. The keys of the return Map are a subset of {@code columns}.
+     */
+    default Map<ColumnMetadata, ViewExpression> getColumnMasks(Identity identity, AccessControlContext context, CatalogSchemaTableName tableName, List<ColumnMetadata> columns)
+    {
+        return Collections.emptyMap();
+    }
+```
+
+ViewExpression class to hold a filter/mask expression
+
+```java
+public ViewExpression(String identity, Optional<String> catalog, Optional<String> schema, String expression)
+```
+
+Analysis.java will hold filters and masks for the table with additional methods
+
+```java
+void registerTableForRowFiltering(QualifiedObjectName table, String identity)
+
+boolean hasRowFilter(QualifiedObjectName table, String identity)
+
+void addRowFilter(Table table, Expression filter)
+
+List<Expression> getRowFilters(Table node)
+
+void registerTableForColumnMasking(QualifiedObjectName table, String column, String identity)
+
+boolean hasColumnMask(QualifiedObjectName table, String column, String identity)
+
+void addColumnMask(Table table, String column, Expression mask)
+
+Map<String, Expression> getColumnMasks(Table table)
+```
+
+#### Additional information
+
+1. What modules are involved
+   2. `presto-main`
+   3. `presto-spi`
+   4. `presto-analyzer` to hold masks and filters
+   5. `presto-hive` for legacy and sql access control
+2. Any new terminologies/concepts/SQL language additions
+   3. NA
+3. Method/class/interface contracts which you deem fit for implementation.
+   4. NA
+4. Code flow using bullet points or pseudo code as applicable
+   5. During analysis phase, access control apis used to retrieve and analyze row filters and column masks.
+   6. Analyzed filters and masks are stored in `Analysis`.
+   7. `RelationPlanner` will the get the filters and masks from `Analysis` and rewrite a new plan with them applied.
+   8. By default no filters or masks are added and the plan will not be rewritten.
+5. Any new user facing metrics that can be shown on CLI or UI.
+   1. NA
+
+## [Optional] Metrics
+
+This is a 0 to 1 feature and will not have any metrics.
```

Comment:
It would be nice to have a metric or field in the query info that tracks how many (a count) row filters/column masks were applied to a TableScan

Reply:
Yes, that sounds like it could be useful. But the access control plugin will know exactly what filters and masks are being applied, so couldn't that be done from within the plugin?

Reply:
So if we have both system and catalog level filters, we would be better off having a system-level metric that capture's all applied filters/masks Also, at the moment, we do not pass/create a `ConnectorSession` which has the `RuntimeStats` to be used for adding metrics

I'm okay with having metrics for this feature as a non-goal; these can be added later